### ISSUE TITLE
Pyston, NumPy, SciPy, Pandas

### ISSUE THREAD

**Issue by pcdinh** [pyston/pyston#634]
Hi,

I don't know if it is too early to ask this question: Can Pyston support NumPy, SciPy, Pandas when it is released or lots of works are required to port NumPy, SciPy, Pandas to Pyston?

PyPy has a major drawbacks when working with those libraries, making it irrelevant for use in scientific computing. See http://cyrille.rossant.net/whats-wrong-with-scientific-python/ or http://www.anlytcs.com/2014/01/how-i-reduced-my-data-crunching-time.html

Thanks,

Dinh

**Comment by kmod** [comment]
Hi Dinh, I'd say there are two sides to the issue -- on the theoretical side, we should have an easier time supporting those libraries since we are more closely adhering to CPython's C API.  On the practical side, those are all very large projects so there's going to be a lot of work regardless.  For exmaple, CPython is pretty forgiving about certain types of bugs, so it's quite easy for extension code to accumulate those bugs.  So we have to spend the time of finding those spots and either fix the extension code or make sure we support the buggy behavior.

And finally there's the issue of prioritization: the first service that we are trying to support doesn't need numpy, so it'd be hard for us to justify investing time into it right now.  But personally I really hope that we can support numpy at some point, since I think it's part of the strength of our approach; we would definitely love if there were contributors interested in making it happen :)  And then maybe down the line we can integrate with something like Numba and get even better performance.

**Comment by rudi-c** [comment]
I did some initial investigation on support `NumPy` and I can get it to a point where it can run some simple linear algebra functions and Mandelbrot calculations without crashing (haven't checked if the results are right). That required commenting out a few things and patching NumPy, but they're either not major (we're missing a non-essential feature like __doc__ on a type) or "might crash in the future but currently doesn't".

Things that need to be done to reduce the amount of commenting stuff out include #775 #776 #777 #778 #779.

**Comment by njsmith** [comment]
FWIW as a numpy dev, I'd definitely be interested in talking about numpy/pyston interoperability stuff at some or any point.

**Comment by rudi-c** [comment]
Thanks @njsmith! The offer is very much appreciated. We are still at the preliminary stages of figuring it out what interoperability work needs to be done. So far we have run into surprisingly few issues (most involving missing code on our end).

If you're curious about the issues that we did run into, here are three that we currently resort to patching:
1) Direct use of `PyStringObject`'s `ob_shash` and `ob_sstate` fields (our `BoxedString` object does not have them).
2) Direct use of the descriptor structs (we currently have our custom structs instead of `PyDescrObject` and related structs -- I haven't looked into if it's easy or not for us to have them).
3) Static Python objects that are not `PyTypeObject`s. This is the more challenging one for us. For static type objects, they are all initialized via some type initialization methods and we can register them as GC roots. However, there is no function that gets called to initialize any Python object that does not get allocated dynamically, so we have no way of registering those objects into our GC. Note that we don't actually need to register them in our GC for the GC to work, thanks to conservative scanning. We just lose the ability to assert on whether an object is valid or not in some places.

**Comment by njsmith** [comment]
Re: (1), I see what you mean... unfortunately not sure if there's much we can we do about it, given the way `np.string_` subclasses `str`, but if you have any ideas then would love to hear.

Re: (2), not sure what you're referring to. Surely you don't mean that Pyston is missing structs like `PyNumberMethods`, that are required by every C-defined type? (Or maybe you could mean that you can use them when defining types, but can't later use `obj->ob_type->tp_as_number->nb_add` to chase down the `nb_add` implementation for an arbitrary object?) -- Or, maybe better than us playing guessing games and going back and forth on github, are the patches you mention available somewhere to look at? :-)

Re: (3): Ugh, not sure where we do this but agreed that static non-`PyTypeObject`-objects is totally illegal in the CPython API; would definitely be interested in patches to get rid of this, or at least a bug report pointing out where we do it.

At a higher level: As you can probably imagine, numpy users are *extremely* interested in the idea of Python JIT support. I think of this as having three important levels, based on what sort of code can reasonably get optimized, where each level is an order of magnitude more compelling to numerical users:

Level 0: code like this is fast:
```
def add3vectors(list_of_floats1, list_of_floats2, list_of_floats3):
    assert len(list_of_floats1) == len(list_of_floats2) == len(list_of_floats3)
    result = [0.0] * len(list_of_floats1)
    for i in range(len(list_of_floats1)):
        result[i] = list_of_floats1[i] + list_of_floats2[i] + list_of_floats3[i]
    return result
```

Level 1: the above code remains fast when using numpy arrays instead of built-in lists

Level 2: this implementation is also fast:
```
def add3vectors(array1, array2, array3):
    return array1 + array2 + array3
```

I assume you have level 0 well in hand :-). What's interesting about the other cases, though, is that performing well on level 1 requires the ability to inline `ndarray.__getitem__` and `ndarray.__setitem__` (at least for simple scalar indexing cases -- Cython and Numba actually handle this by special-casing numpy arrays / objects that implement the memoryview protocol in their compilers), and level 2 requires the ability to inline the two calls to `ndarray.__add__` and then eliminate the temporary and fuse the resulting loops, to produce code like in the examples above. (And interestingly, AFAICT this last case is one that a function-at-a-time JIT in principle ought to be able to handle better than a tracing JIT? Last time I checked PyPy's pseudo-numpy library was totally cheating to handle this last case, because otherwise the tracer had no way to notice that this hot loop it was executing now was right next to another loop over the same data that was going to become hot as soon as this one became cold.) So for the numerical cases I think the key question is: what kind of strategy could we use to expose numpy's internals in a way that's transparent to the JIT? Could we... use clang to compile C Python extensions code to LLVM IR and then hand that to the JIT? ...rewrite key bits of numpy in Python-with-Cython-style-annotations and teach Pyston to JIT this extended Python? ...?

Obviously this is a longer-term conversation, and I know that at this stage you're still focused on getting together an MVP that will satisfy your internal customers. But, I just wanted to put that out there as something that might be worth starting to ponder in the background :-).

[BTW, I'm in [Berkeley](http://bids.berkeley.edu/); I'm guessing you're in the Bay Area too? I'd be happy to chat over coffee sometime, or if you want to meet the crowd at BIDS then you'd be welcome to visit...]

**Comment by rudi-c** [comment]
Sorry for the delay - I was preparing the NumPy patch. It can be seen here: https://github.com/dropbox/pyston-testsuite/commit/3387c959b168c2d6190261af1f99d1c3ab3284c8

The static non-type objects that I found are the ones in the `_builtin_descrs` array in `numpy/core/src/multiarray/arraytypes.c.src` and `_PyArrayScalar_BoolValues` in `numpy/core/src/multiarray/scalartypes.c.src`

As far as the JIT support is concerned, I unfortunately only have cursory familiarity with the Pyston code generation code. @kmod would be the expert here. We're both in San Francisco so pretty close by :).

**Comment by kmod** [comment]
This would definitely be interesting to work on :)  We have a number of features that could help -- we have the ability to inline our C++ runtime into the jitted code and then have the LLVM optimizer do nice things like allocation removal, and presumably we could do that to extension code as well.  You're right though that our focus is elsewhere at the moment since Dropbox's webserver workload is quite a bit different; for example loop performance seems to not matter very much right now.

I don't know much about the domain, but I wonder how much numerical-computing-specific stuff needs to be done to be a competitive option.  For instance, I'm sure we could get to the point that your level 1 is fast, but we would still be doing certain things like checking for signals which would make it tricky to enable vectorization.  I'm sure we could add a "hey don't worry about checking for signals in this region" feature, but I guess that's what I meant about domain-specific work.  Maybe one way this could play out is if we can integrate Numba and leverage the work they've already done :)

Anyway, we are definitely aware of the large Python scientific community and I'd personally be excited to have Pyston be useful there, but practically speaking it'll be a bit before we can think about it in earnest.  We'll keep you posted :)

**Comment by njsmith** [comment]
Huh, the signal checking thing is a really interesting point. I suppose the ideal would be that by default something like those "level 0" and "level 1" loops get compiled into the equivalent of:
```
for (intp i = 0; i < size; ++i) {
    out[i] = in1[i] + in2[i] + in3[i];
    if (unlikely(signal_flag_is_set)) {
        deoptimize_and_handle_signal();
    }
}
```
And this would be pretty fast since `unlikely(signal_flag_is_set)` is an extremely predictable branch (assuming that we can get away without needing expensive memory barriers on every check). I guess the interestingly tricky bit in accomplishing this would be to somehow teach the code generation/optimization engine to enforce the constraint "the signal flag needs to be checked every once in a while, like at least once every 100 operations or whatever, *but it doesn't really matter when exactly when*". Maybe it's not so bad if you can define the "every N operations" constraint in terms of post-optimization IR operations instead of in terms of Python operations. ...Sorry, I'll stop backseat driving from a state of total ignorance about how Pyston works :-)

Beyond that, some "go faster" directives are nice-to-have, but not crucial. The set of interesting ones for numerical purposes is probably not that hard to enumerate -- `no_signal_checking` as per above would be nice-to-have, plus Cython supports a small set for indexing operations:
* `boundscheck=False`: disables overflow checking -- compilers are pretty good at hoisting such checks these days but might still be useful sometimes
* `wraparound=False`: disables support for negative indices like `a[-1]`
* `cdivision=True`: C and Python [disagree](https://github.com/cython/cython/wiki/enhancements-division) about how `int` division works; this tells Cython to generate simpler/faster C code that just uses the C division operator directly. Not sure this is even relevant if you're compiling directly without passing through C.

And of course there are probably other general purpose directives that I can imagine might or might not be useful (`with I_promise_not_to_overwrite_builtins:`...) but you don't need our help to think of those :-).

I think the main thing though is -- as far as what needs to be done to be "competitive" in the numerical community, the answer is: none of these :-). Our main requirements for a MVP (= people will start seriously working with it and possibly investing in it) are: (a) ability to use CPython extensions, b/c there's a huge stack of code built on top of the NumPy+CPython C API, (b) not much slower than CPython, (c) some plausible story for how we could eventually climb up the levels I listed above without having to maintain two separate forks of the whole stack with one targeting CPython and the other targeting <new thing>.

We've mostly given up on PyPy for totally failing at (a) and (c). Numba and Cython are neat because they act as incremental add-ons to CPython, so that takes care of (a) and (b) automatically, but they still don't really have any story on (c) -- they're just incrementally reimplementing more and more of NumPy internally. Plus the idea of having a single runtime automagically optimizing your entire program is very appealing. So that's why Pyston's interesting to us :-)

**Comment by njsmith** [comment]
Oh, another point I just remembered: the other interesting thing about making "level 2" fast is that numpy uses the Python type system in a slightly weird way that has implications for speculation.

Thinking of C++ for a moment for an analogy: if you have a function `vector_add(std::vector<T> a, std::vector<T> b)`, you don't want to generate a single copy of the function that handles all `T`; you want to generate one instantiation for each possible `T`. In Python, `ndarray` is effectively a parametrized type like `std::vector`, but in a runtime dispatch world. So instead of a formal type argument `T`, ndarrays have a special `.dtype` attribute, which is sort of like a Python class object except designed for unboxed types instead of boxed types. Operations like `ndarray.__add__` have multiple implementations which they dispatch to depending on whether `a.dtype` says this is an array of ints or floats or whatever. So basically, this means that for ndarrays it's not enough to speculate on the Python-level type of the input arguments; you also want to speculate that the argument will be ndarray-of-int or ndarray-of-float or whatever.

Once you've done this, then in principle the rest can "just work" -- you inline `ndarray.__add__`, then the dispatch logic it contains can be evaluated out at compile time, etc. But you do need some way to know that the dtype is a good thing to speculate on.

(I believe that Numba handles this with a hard-coded heuristic that treats ndarray+dtype as a unit for speculation purposes, and in Cython of course you have to provide the fully parametrized static type ndarray[int64] or whatever if you want faster-than-Python code.)

**Comment by datnamer** [comment]
Really exciting conversation! May be easier/ better idea  to target python's next gen array lib, [Dynd](https://github.com/libdynd/dynd-python)  It has huge improvements  over numpy like native missing data support, variable length strings, more extensible type system etc

Also planning to have compat with numpy code using buffer protocol.

Perhaps core devs @izaid and @mwiebe can chime in here.

**Comment by kmod** [comment]
I doubt we'd wade into any sort of numpy-vs-dynd debate, but if all goes well we should be able to serve both projects :)

**Comment by izaid** [comment]
Hey guys. Guess I'm here to represent DyND now anyway. :)

Of course, we'd be interested in supporting Pyston, as well as other JITs. We actually do have experimental interoperability with Numba already. We'd certainly be keen to explore Pyston over the next few months.
